[ARM] Allow FP reg conversion when copying Sx to Dx

eleviant opened the pull request:
This allows copying float value to Dx reg using inline asm, e.g:
```
float a;
void b() {
  register float d1 asm("d1") =
      a;
  asm("" ::"r"(d1));
}
```

efriedma-quic commented:
Why do you think the compiler should do this?  It's not what gcc does.  And in general implicitly performing floating-point conversions seems like a bad idea.

eleviant commented:
> Why do you think the compiler should do this? It's not what gcc does. And in general implicitly performing floating-point conversions seems like a bad idea.

Well, the inline assembly in summary actually instructs to put float value into 64-bit FP reg and that's exactly what I did. GCC seems to put value into s1 instead of d1, which seems confusing. Compiling the same with clang results in undefined behavior in release build w/o asserts, (for me it consumes all available memory and crashes). What's the sane thing to do here in your opinion? Emit an error and exit? Doing what GCC does?

efriedma-quic commented:
There are two things wrong in the original testcase:

- `asm("d1")` doesn't match the size of "float".
- `asm("d1")` doesn't match `"r"(d1)`; it's using the wrong kind of register.  The clang frontend currently throws away the register specification, though, so it's not relevant to the backend.  (Maybe we should diagnose in clang.)

-----

An error seems fine for a mismatch like this  Doing the same thing as gcc is also probably okay (putting the float value the lower half of d1, which is s2).  In any case, we probably should handle this during SelectionDAG, in ARMTargetLowering::getRegForInlineAsmConstraint or something like that.

eleviant committed:
[ARM] Emit error message when incompatible reg is specified

At the moment the following piece of code casues undefined behaviour:
```
int a;
void b() {
  register float d2 asm("d2") =
      a;
  asm("" ::"r"(d2));
}
```
This happens because variable and register types are incompatible.

efriedma-quic commented on llvm/lib/Target/ARM/ARMISelLowering.cpp:
```diff
@@ -20357,6 +20357,13 @@ ARMTargetLowering::getSingleConstraintMatchWeight(
+static bool isIncompatibleReg(const MCPhysReg &PR, MVT VT) {
+  if (VT == MVT::Other)
+    return false;
+  return (ARM::SPRRegClass.contains(PR) && VT != MVT::f32) ||
```
```suggestion
  return (ARM::SPRRegClass.contains(PR) && VT != MVT::f32 && VT != MVT::i32) ||
```

We currently accept i32/f32 mismatches, and do the appropriate bitcast.  I think it makes sense to continue doing that.

The other combinations have never worked, as far as I can tell, so probably fine to reject.

eleviant commented:
@efriedma-quic PTAL

efriedma-quic reviewed (commented):
One minor comment on the test, otherwise looks fine.

eleviant commented:
@efriedma-quic Addressed review comments

efriedma-quic reviewed (approved):
LGTM

mstorsjo commented:
This change seems to break cases that seem entirely legit (although slightly unusual) to me:
```c
#include <arm_neon.h>
void qt_convert_rgb888_to_rgb32_neon() {
  register uint8x8_t fullVector asm("d3");
  asm("" : : "w"(fullVector));
}
```
Here, `uint8x8_t` and the `w` constraint should be the right ones for the register `d3`.

eleviant commented:
@mstorsjo See https://github.com/llvm/llvm-project/pull/152175

nikic commented:
Does this intentionally forbid using s reg constraints for half float? Don't those use s registers under the hard float ABI?
